### PR TITLE
[query] improve asm4s, Emit codegen

### PR DESCRIPTION
**Pull request by cseed**
Large collection of interface improvements to asm4s and Emit {Emit}{Module, Class, Method, Function}Builder.  The main goal here is to make generating arbitrary classes a first class activity.  Here is a summary of changes:
 - newLocal now has parameters: newLocal().  This is a step towards requiring names on all generated objects.
 - {Emit}{Class, Method, Function}Builder now takes a type parameter that represents (a supertype) of the class being built: e.g. MethodBuilder[C] is a builder for a method on a class of type C.  Note, we can't have a type parameter that represents the actual class type because that doesn't exist yet.
 - {Emit}FunctionBuilder all but gone: {Emit}FunctionBuilder is now just a {Emit}MethodBuilder is an apply method.  Most functionality moved to {Emit}ClassBuilder.
 - Added EmitClassBuilder.
 - It is convenient to have e.g. MethodBuilder support the ClassBuilder interface: this is what the Wrapped traits are for: MethodBuilder extends WrappedClassBuilder and ClassBuilder extends WrappedModuleBuilder.  So MethodBuilder has the ClassBuilder interface, but is not actually a ClassBuilder.  I tried a bunch of variants for the design of this, and while I don't think this is quite perfect, it seems workable.
 - EmitMethodBuilder extends WrappedMethodBuilder, etc.  Rather than overloading, the two interfaces are distinct: genMethod vs genEmitMethod, etc.
 - Pushed "new vs gen" into more places e.g. newMethod vs genMethod.  newMethod takes a name and creates a method of that name (e.g. apply).  genMethod takes a baseName and creates a unique name based on the baseName.
 - MethodBuilder newField => genFieldThisRef to distinguish it from ClassBuilder.newField.  The former returns a Settable[T] referencing `this.<field>`, the latter just returns a Field.
 - All methods supporting code generation for IR take EmitMethodBuilder rather than MethodBuilder (PType routines, aggregators, etc.)

Summarizing the new class structure:

```
class ModuleBuilder
trait WrappedModuleBuilder
  def modb: ModuleBuilder
class ClassBuilder[C] extends WrappedModuleBuilder
trait WrappedClassBuilder[C]
  def cb: ClassBuilder[C]
class MethodBuilder[C] extends WrappedClassBuilder[C]
trait WrappedMethodBuilder[C]
  def mb: MethodBuilder
class FunctionBuilder[F] extends WrappedMethodBuilder[F]
  def apply_method: MethodBuilder[F]

class EmitModuleBuilder extends WrappedModuleBuilder
trait WrappedEmitModuleBuilder extends WrappedModuleBuilder
  def emodb: EmitModuleBuilder
class EmitClassBuilder[C] extends WrappedEmitModuleBuilder with WrappedClassBuilder[C]
trait WrappedEmitClassBuilder[C] extends WrappedModuleBuilder with WrappedClassBuilder[C]
  def ecb: EmitClassBuilder
class EmitMethodBuilder[C] extends WrappedClassBuilder[C] with WrappedMethodBuilder[C]
trait WrappedEmitMethodBuilder[C] extends WrappedClassBuilder[C] with WrappedMethodBuilder[C]
  def emb: EmitMethodBuilder
class EmitFunctionBuilder[F] extends WrappedEmitMethodBuilder[F]
  def apply_method: EmitFunctionBuilder[F]
```

### CODE REVIEW

**Comment by cseed** [comment]
I pushed another commit to make registered function names valid Java identifiers because I'm using the function name when I emit a Java method for a registered function.

**Comment by cseed** [comment]
I think I squashed all the bugs, failure is from the unrelated test_ci issue.

File: hail/src/test/scala/is/hail/expr/ir/IntervalSuite.scala [hunk 1]
```diff
@@ -164,6 +164,7 @@ class IntervalSuite extends HailSuite {
     ))
 
     assertEvalsTo(invoke("sortedNonOverlappingIntervalsContain", TBoolean, intervals, I32(-1)), false)
+    /*
```
Comment (tpoterba) [inline hail/src/test/scala/is/hail/expr/ir/IntervalSuite.scala]: ?

Comment (cseed) [reply hail/src/test/scala/is/hail/expr/ir/IntervalSuite.scala]: Oops, fixed.

**Pull request merged**